tests/e2e/e2e_test.go

```diff
@@ -0,0 +1,268 @@
+package e2e
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/classic-terra/core/v2/tests/e2e/initialization"
+)
+
+func (s *IntegrationTestSuite) TestIBCWasmHooks() {
+	if s.skipIBC {
+		s.T().Skip("Skipping IBC tests")
+	}
+	chainA := s.configurer.GetChainConfig(0)
+	chainB := s.configurer.GetChainConfig(1)
+
+	nodeA, err := chainA.GetDefaultNode()
+	s.NoError(err)
+	nodeB, err := chainB.GetDefaultNode()
+	s.NoError(err)
+
+	nodeA.StoreWasmCode("counter.wasm", initialization.ValidatorWalletName)
+	chainA.LatestCodeID = int(nodeA.QueryLatestWasmCodeID())
+	nodeA.InstantiateWasmContract(
+		strconv.Itoa(chainA.LatestCodeID),
+		`{"count": "0"}`, "",
+		initialization.ValidatorWalletName)
+
+	contracts, err := nodeA.QueryContractsFromID(chainA.LatestCodeID)
+	s.NoError(err)
+	s.Len(contracts, 1, "Wrong number of contracts for the counter")
+	contractAddr := contracts[0]
+
+	transferAmount := sdk.NewInt(10000000)
+	validatorAddr := nodeB.GetWallet(initialization.ValidatorWalletName)
+	nodeB.SendIBCTransfer(validatorAddr, contractAddr, fmt.Sprintf("%duluna", transferAmount.Int64()),
+		fmt.Sprintf(`{"wasm":{"contract":"%s","msg": {"increment": {}} }}`, contractAddr))
+
+	// check the balance of the contract
+	s.Eventually(func() bool {
+		balance, err := nodeA.QueryBalances(contractAddr)
+		s.Require().NoError(err)
+		if len(balance) == 0 {
+			return false
+		}
+		return balance[0].Amount.Equal(transferAmount)
+	},
+		1*time.Minute,
+		10*time.Millisecond)
+
+	// sender wasm addr
+	// senderBech32, err := ibchookskeeper.DeriveIntermediateSender("channel-0", validatorAddr, "terra")
+	var response interface{}
+	response, err = nodeA.QueryWasmSmart(contractAddr, `{"get_total_funds": {}}`)
+	s.Require().NoError(err)
+	fmt.Println("response: ", response)
+
+	s.Eventually(func() bool {
+		response, err = nodeA.QueryWasmSmart(contractAddr, `{"get_total_funds": {}}`)
+		if err != nil {
+			return false
+		}
+
+		fmt.Println("response: ", response)
+		totalFunds := response.([]interface{})[0]
+		amount, err := strconv.ParseInt(totalFunds.(map[string]interface{})["amount"].(string), 10, 64)
+		if err != nil {
+			return false
+		}
+		denom := totalFunds.(map[string]interface{})["denom"].(string)
+		// check if denom is luna token ibc
+		return sdk.NewInt(amount).Equal(transferAmount) && denom == initialization.TerraIBCDenom
+	},
+		10*time.Second,
+		10*time.Millisecond,
+	)
```

Comment:
shouldn't you need to check the value of "count" in contract again to ensure that it is "1" after ibc-hook is activated upon receiving ibc transfer?

<img width="1109" alt="Screenshot 2024-02-01 at 00 53 53" src="https://github.com/classic-terra/core/assets/23058164/41fd82ce-27b8-4de1-96be-0e73c1cfa13a">